File: 백준/Gold/16234. 인구 이동/인구 이동.py
import sys
from collections import deque
input = sys.stdin.readline

n, l, r = map(int, input().split())

states = []
for _ in range(n):
    states.append(list(map(int, input().split())))


dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def bfs(a,b):
    queue = deque()
    temp = []
    queue.append((a,b))
    temp.append((a,b))
    while queue:
        x,y = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0<=nx<n and 0<=ny<n and not visited[nx][ny]:
                if l<=abs(states[nx][ny]-states[x][y])<=r:
                    visited[nx][ny] = 1
                    queue.append((nx,ny))
                    temp.append((nx,ny))
    return temp

ans = 0
while True:
    is_moved = False
    visited = [[False] * (n) for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if not visited[i][j]:
                visited[i][j] = True
                united = bfs(i,j)
                if len(united) > 1:
                    is_moved = True
                    united_avg = sum(states[a][b] for a,b in united) // len(united)
                    for a,b in united:
                        states[a][b] = united_avg
    if not is_moved:
        break
    ans += 1

print(ans)

File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.io.*;
import java.util.*;
import java.math.*;

class Pair {
    int x, y;

    Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

public class Main {
    static int n, l, r;
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static int[][] states;
    static boolean[][] visited;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        l = Integer.parseInt(st.nextToken());
        r = Integer.parseInt(st.nextToken());

        states = new int[n][n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                states[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int ans = 0;
        while (true) {
            boolean isMoved = false;
            visited = new boolean[n][n];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (!visited[i][j]) {
                        visited[i][j] = true;
                        List<Pair> united = bfs(i, j);
                        if (united.size() > 1) {
                            isMoved = true;
                            int unitedSum = 0;
                            for (Pair pair : united) {
                                unitedSum += states[pair.x][pair.y];
                            }
                            int unitedAvg = unitedSum / united.size();
                            for (Pair pair : united) {
                                states[pair.x][pair.y] = unitedAvg;
                            }
                        }
                    }
                }
            }
            if (!isMoved) {
                break;
            }
            ans++;
        }
        
        bw.write(ans + "\n");
        bw.flush();

        bw.close();
        br.close();
    }

    static List<Pair> bfs(int a, int b) {
        List<Pair> temp = new ArrayList<>();
        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(a, b));
        temp.add(new Pair(a, b));
        while (!queue.isEmpty()) {
            Pair current = queue.poll();
            for (int i = 0; i < 4; i++) {
                int nx = current.x + dx[i];
                int ny = current.y + dy[i];
                if (0 <= nx && nx < n && 0 <= ny && ny < n && !visited[nx][ny]) {
                    if (l <= Math.abs(states[nx][ny] - states[current.x][current.y]) && Math.abs(states[nx][ny] - states[current.x][current.y]) <= r) {
                        visited[nx][ny] = true;
                        queue.add(new Pair(nx, ny));
                        temp.add(new Pair(nx, ny));
                    }
                }
            }
        }
        return temp;
    }
}